internal/aws/k8s/k8sclient/job.go

```diff
@@ -0,0 +1,190 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8sclient
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	batchv1 "k8s.io/api/batch/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	CronJob = "CronJob"
+)
+
+type JobClient interface {
+	JobToCronJob() map[string]string
+
+	Init()
+	Shutdown()
+}
+
+type jobClient struct {
+	sync.RWMutex
+
+	stopChan chan struct{}
+
+	store *ObjStore
+
+	reflector cacheReflector
+
+	inited bool
+
+	pollInterval time.Duration
+	pollTimeout  time.Duration
+
+	cachedJobMap    map[string]time.Time
+	jobToCronJobMap map[string]string
+}
+
+func (c *jobClient) JobToCronJob() map[string]string {
+	if !c.inited {
+		c.Init()
+	}
+	if c.store.Refreshed() {
+		c.refresh()
+	}
+	c.RLock()
+	defer c.RUnlock()
+	return c.jobToCronJobMap
+}
+
+func (c *jobClient) refresh() {
+	c.Lock()
+	defer c.Unlock()
+
+	objsList := c.store.List()
+
+	tmpMap := make(map[string]string)
+	for _, obj := range objsList {
+		job := obj.(*jobInfo)
+	ownerLoop:
+		for _, owner := range job.owners {
+			if owner.kind == CronJob && owner.name != "" {
+				tmpMap[job.name] = owner.name
+				break ownerLoop
+			}
+		}
+	}
+
+	if c.jobToCronJobMap == nil {
+		c.jobToCronJobMap = make(map[string]string)
+	}
+
+	if c.cachedJobMap == nil {
+		c.cachedJobMap = make(map[string]time.Time)
+	}
+
+	lastRefreshTime := time.Now()
+
+	for k, v := range c.cachedJobMap {
+		if lastRefreshTime.Sub(v) > cacheTTL {
+			delete(c.jobToCronJobMap, k)
+			delete(c.cachedJobMap, k)
+		}
+	}
+
+	for k, v := range tmpMap {
+		c.jobToCronJobMap[k] = v
+		c.cachedJobMap[k] = lastRefreshTime
+	}
+}
+
+func (c *jobClient) Init() {
+	c.Lock()
+	defer c.Unlock()
+	if c.inited {
+		return
+	}
+
+	ctx := context.Background()
+	if _, err := Get().ClientSet.BatchV1().Jobs(metav1.NamespaceAll).List(ctx, metav1.ListOptions{}); err != nil {
+		panic(fmt.Sprintf("Cannot list Job. err: %v", err))
+	}
+
+	c.stopChan = make(chan struct{})
+
+	c.store = NewObjStore(transformFuncJob)
+
+	lw := createJobListWatch(Get().ClientSet, metav1.NamespaceAll)
+
+	//reflector is only set in tests
+	if c.reflector == nil {
+		c.reflector = cache.NewReflector(lw, &batchv1.Job{}, c.store, 0)
+	}
+	go c.reflector.Run(c.stopChan)
+
+	//pollInterval and pollTimeout are only set in tests
+	if c.pollInterval == 0 || c.pollTimeout == 0 {
+		c.pollInterval = 50 * time.Millisecond
+		c.pollTimeout = 2 * time.Second
+	}
+	if err := wait.Poll(c.pollInterval, c.pollTimeout, func() (done bool, err error) {
+		return c.reflector.LastSyncResourceVersion() != "", nil
```

Comment:
This condition seems to be duplicated across many files, can the abstraction be changed so there is only one thing managing initialization of the reflector, not each part?